File: app/domains/engagement/routers/badges.py
"""成就徽章墙（新功能 B 升级后：本文件退化为薄 HTTP 层）

规则与授予逻辑已下沉到 `app/domains/engagement/services/achievement.py`
（单一真相源 BADGE_RULES + 事件驱动 try_grant）。此处只负责：
- 组装 HTTP 响应（含每枚徽章的 progress 进度与 category 分类，供前端进度条/分类 tab）；
- 保持 `GET /api/badges` 的**兜底全量授予**语义（登录/进入徽章墙时补齐历史达标徽章）。
实时授予由各写操作埋点调用 `AchievementService.try_grant`（见 engagement/contracts.py）。
"""
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.domains.engagement.services.achievement import list_badges

router = APIRouter(tags=["badges"])


@router.get("", summary="成就徽章墙：扫描并授予新徽章，返回全部徽章状态（含进度/分类）")
def get_badges(user_id: str = Query(...), db: Session = Depends(get_db)):
    """成就徽章墙：兜底授予新徽章，返回全部徽章状态与进度。

    查询参数：user_id；无需家长密码。
    返回：{total, earned(已得数), newly(本次新解锁 code 列表),
          categories:[{key,label,total,earned}],
          items:[{code,emoji,name,desc,category,earned,earned_at,
                  progress:{current,target,pct}}]}。
    副作用：一次性统计学习数据，对首次达成阈值且未记录的徽章新增 badge_earned 落库
          （阈值与分类见 services/achievement.py 的 BADGE_RULES）。
    """
    return list_badges(db, user_id)

File: web/src/logic/badges.js
// badges.js：成就徽章墙（新功能 B 升级：进度条 + 分类）的 data / computed / methods。
// 原徽章逻辑与「知识卡图鉴」混在 logic/cards.js，本次按功能拆出独立模块。
// 仅依赖通用 api/showToast（留在主文件）；与其他 logic/* 经展开运算符合并后 this 仍绑定同一实例。
//
// 后端契约（app/domains/engagement/services/achievement.py + routers/badges.py）：
//   GET /api/badges?user_id=  ->
//     { total, earned, newly:[code],
//       categories:[{key,label,total,earned}],
//       items:[{code,emoji,name,desc,category,earned,earned_at,progress:{current,target,pct}}] }
// 实时授予由后端写操作埋点完成（交卷 / 掌握错题 / 完成任务 / 心情打卡），前端无需轮询；
// 每枚徽章的 progress 由后端「指标当前值 / 阈值」派生，前端只负责渲染进度条。

export function badgesData() {
  return {
    badgeData: null,   // GET /api/badges 的完整响应（含 items/categories/total/earned/newly）
    badgeNew: [],      // 本次新解锁的徽章（顶部「恭喜获得」卡片 + Toast 用）
    badgeCat: '',      // 分类过滤：'' = 全部，否则为 categories[].key（study/persist/social）
  };
}

export const badgesComputed = {
  // 当前分类下要渲染的徽章列表（'' 显示全部，保持后端返回顺序）
  badgeItems() {
    const d = this.badgeData;
    if (!d || !d.items) return [];
    if (!this.badgeCat) return d.items;
    return d.items.filter(b => b.category === this.badgeCat);
  },
  // 分类 tab 数据：在「全部」聚合项后接后端 categories
  badgeCats() {
    const d = this.badgeData;
    if (!d || !d.categories) return [];
    return [{ key: '', label: '全部', total: d.total, earned: d.earned }].concat(d.categories);
  },
};

export const badgesMethods = {
  /* ─────────── 成就徽章墙（新功能 B：实时授予 + 进度 + 分类） ─────────── */
  // 拉取徽章墙。announce=true 时对新解锁徽章弹 Toast（进入徽章页时用）
  loadBadges(announce) {
    if (!this.user) return;
    this.api(`/api/badges?user_id=${encodeURIComponent(this.user)}`)
      .then(d => {
        const prevNew = this.badgeNew;
        this.badgeData = d;
        if (d.newly && d.newly.length) {
          this.badgeNew = (d.items || []).filter(b => (d.newly || []).includes(b.code));
          if (announce && this.badgeNew.length) {
            this.showToast(`🎉 获得新徽章：${this.badgeNew.map(b => b.name).join('、')}！`);
          }
        } else if (prevNew.length) {
          this.badgeNew = prevNew;
        }
      })
      .catch(() => { this.badgeData = null; });
  },
  // 切换分类过滤（前端本地过滤，不重新请求）
  setBadgeCat(key) {
    if (this.badgeCat === key) return;
    this.badgeCat = key;
  },
};
